### PR TITLE
Fix invalid expect block and support for multi-line errors

### LINKED ISSUE #58: In REPL, if you can determine that an error has occurred on one line, you should not start a new line
image

Here it should probably just report the error instead of starting a new line

### PR DESCRIPTION
**Pull request by GreasySlug**
Fixes #58  and #248.

Each line of the REPL is now evaluated by the `Parser::eval()` (Parser and Lex are not modified)

Changes proposed in this PR:
- Add Runnable Parser to check invalid syntax
- Check invalid block as a error
- Add block_begin to use for REPL
- Display multi-line error

@mtshiba

### CODE REVIEW

File: compiler/erg_common/error.rs [hunk 1]
```diff
@@ -45,6 +45,7 @@ pub enum ErrorKind {
     VisibilityError = 20,
     MethodError = 21,
     DummyError = 22,
+    ExpectNewLine = 23,
```
Comment (GreasySlug) [inline compiler/erg_common/error.rs:48]: I've tentatively added the error numbers.
Please comment if the right number and location is available.

**Comment by GreasySlug** [comment]
Currently it is implemented only in `dummy.rs`.
Is it necessary to implement `run()` other than `dummy.rs`?

**Comment by mtshiba** [comment]
A method should be added to the `Runnable` trait as a request method to determine if the next line should be followed. The name should be something like `expect_nextline()`.

File: src/dummy.rs [hunk 2]
```diff
@@ -176,6 +180,108 @@ impl Runnable for DummyVM {
...
+                if !quiet_repl {
+                    log!(info_f output, "The REPL has started.\n");
+                    output
+                        .write_all(instance.start_message().as_bytes())
+                        .unwrap();
+                }
+                output.write_all(instance.ps1().as_bytes()).unwrap();
+                output.flush().unwrap();
+                let mut in_block = false;
+                let mut parser = ParserRunner::new(cfg);
```
Comment (GreasySlug) [inline src/dummy.rs:200]: parser is declared here, and I am using it, but it is not possible to add this to the Runnable trait.
Or is there a way around this?

Comment (mtshiba) [reply src/dummy.rs:200]: `expect_nextline` should be declared with no body, and the specific judgment algorithm should be defined in each struct. Compiler, Parser etc. uses Parser, Lexer should simply return `false`.

Comment (GreasySlug) [reply src/dummy.rs:200]: Thank you!!
I'll try it

File: compiler/erg_common/error.rs (hunk 1 above)
Comment (mtshiba) [reply compiler/erg_common/error.rs:48]: 👍

**Comment by GreasySlug** [comment]
I fixed conflict.

**Comment by GreasySlug** [comment]
multi-line-error

File: compiler/erg_common/traits.rs [hunk 3]
```diff
@@ -347,37 +347,38 @@ pub trait LimitedDisplay {
...
-        || src.ends_with("->")
-        || src.ends_with("=>")
-        // when `"""` are on the same line
-        // e.g. """something"""
-        || src.contains("\"\"\"") && src.find("\"\"\"") == src.rfind("\"\"\"")
+#[derive(Debug, PartialEq, Eq)]
+pub enum BlockKind {
+    Main,          // now_block Vec must contain this
+    Assignment,    // =
+    Lambda,        // =>, ->, do:, do!:
```
Comment (GreasySlug) [inline compiler/erg_common/traits.rs:354]: I have included the `:colon` in the Lambda
Is there a better way to handle this?

File: compiler/erg_parser/parse.rs [hunk 4]
```diff
@@ -905,14 +926,24 @@ impl Parser {
         };
         if self.cur_is(Colon) {
             self.lpop();
-            let body = self.try_reduce_block().map_err(|_| self.stack_dec())?;
+            let body = self.try_reduce_block().map_err(|_| {
+                self.stack_dec();
+                if let Some(err) = self.errs.last_mut() {
+                    err.set_hint("Lambda");
```
Comment (GreasySlug) [inline compiler/erg_parser/parse.rs:957]: When an error occurs the last element is available for extraction
Extract it and have a string as a hint

File: compiler/erg_common/traits.rs (hunk 3 above)
Comment (mtshiba) [reply compiler/erg_common/traits.rs:354]: you mean like this?

```python
match! x:
    ...
```

Defining an additional enum option with a name like `ColonCall` may be better.

Comment (GreasySlug) [reply compiler/erg_common/traits.rs:354]: Thank you

**Comment by GreasySlug** [comment]
I fixed #281 then I found some bugs.
so I need more time to test and fix them

**Comment by mtshiba** [comment]
When REPL detects an error, it refers to the previous source information.

スクリーンショット 2024-02-28 20 07 50

**Comment by GreasySlug** [comment]
I addressed the issue occurring during single-line errors.

Additionally, I added error handling for cases where newline are expected, such as with `do!:`, `do:`, to ensure evaluation in the REPL.

error

**Comment by mtshiba** [comment]
REPL references previous source information when dealing with multiple lines of code.

スクリーンショット 2024-02-29 13 30 07

スクリーンショット 2024-02-29 13 31 46

**Comment by GreasySlug** [comment]
- Multi line errors
error

- Collections
collections

**Review by mtshiba — APPROVED** [review]

**Comment by mtshiba** [comment]
Thank you!

**Pull request merged**